# Analisis de datos
## Introducción a Pandas
## Índice:
1. [Extracción de datos](#Extraer-datos:-Restaurant)
2. [Filtrar datos](#Filtrar-datos)
3. [GroupBy](#GroupBy)
4. [Apply](#Apply:-Student-Alcohol-Consumption)
5. [Merge](#Merge:-Housing-Market)
6. [Stats](#Stats:-US---Baby-Names)
7. [Deleting](#Deleting:-Iris)
8. [Bonus](#lab#Bonus:)

<b>Nota:</b>
Esta practica esta basada en los ejercicios de varios recursos online como:

* https://github.com/ajcr/100-pandas-puzzles/blob/master/100-pandas-puzzles.ipynb
* https://www.machinelearningplus.com/python/101-pandas-exercises-python/
* https://www.kaggle.com/learn/pandas
* https://github.com/facuzeta/pandas_exercises
* https://github.com/guipsamora/pandas_exercises

# Extraer datos: Restaurant
Referencias:
* [Read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)
* [Group by](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html)
* [dtype](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html)


1. Import the necessary libraries

In [2]:
import pandas as pd
import numpy as np

2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [29]:
url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv'

## tipos de extensiones típicas datos
# csv -> separados por coma
# tsv -> separados por tabulaciones
# lsv -> separados por espacios
# *csv -> csv importado desde excel puede tener ; de separador
# txt -> puede ser cualquier cosa, suele ser coma


chipo = pd.read_csv(url, delimiter = '\t') # por defecto delimiter = ','

chipo.head()

order_id  quantity                              item_name   
0         1         1           Chips and Fresh Tomato Salsa  \
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

3. Assign it to a variable called chipo.

In [30]:
chipo = chipo

4. See the first 10 entries

In [31]:
chipo.head(10)

order_id  quantity                              item_name   
0         1         1           Chips and Fresh Tomato Salsa  \
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

5. What is the number of observations in the dataset?

In [32]:
len(chipo)

4622

6. What is the number of columns in the dataset?

In [33]:
len(chipo.columns)

5

7. Print the name of all the columns.

In [34]:
chipo.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

9. Which was the most-ordered item? 

In [35]:
chipo.item_name.value_counts()

item_name
Chicken Bowl                             726
Chicken Burrito                          553
Chips and Guacamole                      479
Steak Burrito                            368
Canned Soft Drink                        301
Steak Bowl                               211
Chips                                    211
Bottled Water                            162
Chicken Soft Tacos                       115
Chips and Fresh Tomato Salsa             110
Chicken Salad Bowl                       110
Canned Soda                              104
Side of Chips                            101
Veggie Burrito                            95
Barbacoa Burrito                          91
Veggie Bowl                               85
Carnitas Bowl                             68
Barbacoa Bowl                             66
Carnitas Burrito                          59
Steak Soft Tacos                          55
6 Pack Soft Drink                         54
Chips and Tomatillo Red Chili Salsa       48


10. For the most-ordered item, how many items were ordered?

In [36]:
chipo.loc[chipo['item_name'] == 'Chicken Bowl']['quantity'].sum()


761

11. What was the most ordered item in the choice_description column?

In [37]:
chipo.choice_description.value_counts()

choice_description
[Diet Coke]                                                                                                                                      134
[Coke]                                                                                                                                           123
[Sprite]                                                                                                                                          77
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]                                                                            42
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Guacamole, Lettuce]]                                                                 40
                                                                                                                                                ... 
[Fresh Tomato Salsa (Mild), [Pinto Beans, Black Beans, Rice, Cheese, Sour Cream, Lettuc

12. How many items were orderd in total?

In [38]:
chipo['quantity'].sum()

4972

13. Turn the item price into a float

In [39]:
# chipo['item_price'].str.replace('$', '', regex=True).replace(' ', '', regex=True).astype(float)
chipo['item_price'] = chipo['item_price'].str.replace('$', '')
chipo['item_price'] = chipo['item_price'].str.strip()
chipo['item_price'] = chipo['item_price'].astype(float)

13.a. Check the item price type

In [40]:
chipo['item_price'].dtype

dtype('float64')

13.b. Change the type of the column into a float

In [41]:
chipo['item_price'] = pd.to_numeric(chipo['item_price'])

13.c. Check the item price type

In [42]:
chipo['item_price'].dtype

dtype('float64')

14. How much was the revenue for the period in the dataset?

In [43]:
result = chipo['item_price'] * chipo['quantity']
sum(result)

39237.020000000055

15. How many orders were made in the period?

In [44]:
chipo['order_id'].count()

4622

16. What is the average revenue amount per order?

In [45]:
chipo['item_price'].mean()

7.464335785374297

17. How many different items are sold?

In [46]:
chipo['item_name'].nunique()

50

18. How many products cost more than $10.00?

In [47]:
chipo.loc[chipo['item_price'] >= 10]['item_name'].nunique()

31

# Filtrar datos
Referencias:
* [Reset Index](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html)
* [Sort values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html?highlight=sort_values#pandas.DataFrame.sort_values)

19. What is the price of each item? 
> print a data frame with only two columns item_name and item_price

In [48]:
chipo[['item_name', 'item_price']].drop_duplicates()

item_name  item_price
0              Chips and Fresh Tomato Salsa        2.39
1                                      Izze        3.39
2                          Nantucket Nectar        3.39
3     Chips and Tomatillo-Green Chili Salsa        2.39
4                              Chicken Bowl       16.98
...                                     ...         ...
4237                    Chips and Guacamole        8.50
4354                       Steak Soft Tacos       18.50
4489                    Chips and Guacamole       17.80
4509                                  Chips        1.99
4510                          Barbacoa Bowl       11.49

[209 rows x 2 columns]

20. Sort by the name of the item

In [49]:
chipo[['item_name', 'item_price']].sort_values(by=['item_name']).drop_duplicates()

item_name  item_price
3389  6 Pack Soft Drink       12.98
341   6 Pack Soft Drink        6.49
418       Barbacoa Bowl        9.25
4485      Barbacoa Bowl       11.75
1383      Barbacoa Bowl       11.48
...                 ...         ...
2223  Veggie Salad Bowl        8.75
738   Veggie Soft Tacos       11.25
3889  Veggie Soft Tacos       16.98
2384  Veggie Soft Tacos        8.75
2851  Veggie Soft Tacos        8.49

[209 rows x 2 columns]

21. What was the quantity of the most expensive item ordered?

In [50]:
chipo.loc[chipo['item_name'] == chipo.loc[chipo['item_price'] == chipo['item_price'].max()]['item_name'].item()]['quantity'].count()

110

22. How many times was a Veggie Salad Bowl ordered?

In [51]:
chipo.loc[chipo['item_name'] == 'Veggie Salad Bowl']['item_name'].count()

18

23. How many times did someone order more than one Canned Soda?

In [52]:
chipo.loc[(chipo['item_name'] == 'Canned Soda') & (chipo['quantity'] > 1)]['order_id'].count()

20

# GroupBy

Referencias:
* [Describe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html)

GroupBy can be summarized as Split-Apply-Combine. Check out this [Diagram](http://i.imgur.com/yjNkiwL.png)  

24. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/drinks.csv). 

In [53]:
url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/drinks.csv'

## tipos de extensiones típicas datos
# csv -> separados por coma
# tsv -> separados por tabulaciones
# lsv -> separados por espacios
# *csv -> csv importado desde excel puede tener ; de separador
# txt -> puede ser cualquier cosa, suele ser coma


drinks = pd.read_csv(url, delimiter = ',') # por defecto delimiter = ','

drinks.head()

country  beer_servings  spirit_servings  wine_servings   
0  Afghanistan              0                0              0  \
1      Albania             89              132             54   
2      Algeria             25                0             14   
3      Andorra            245              138            312   
4       Angola            217               57             45   

   total_litres_of_pure_alcohol continent  
0                           0.0        AS  
1                           4.9        EU  
2                           0.7        AF  
3                          12.4        EU  
4                           5.9        AF

25. Assign it to a variable called drinks.

26. Which continent drinks more beer on average?

In [54]:
drinks.groupby('continent').agg(
    average=('beer_servings', np.average)
).sort_values(by=['average'], ascending=False).head(1)


average
continent            
EU         193.777778

27. Print the mean alcohol consumption per continent for every column

In [55]:
drinks.groupby('continent').agg(
    mean_beer=('beer_servings', np.mean),
    mean_spirit=('spirit_servings', np.mean),
    mean_wine=('wine_servings', np.mean),
)

mean_beer  mean_spirit   mean_wine
continent                                     
AF          61.471698    16.339623   16.264151
AS          37.045455    60.840909    9.068182
EU         193.777778   132.555556  142.222222
OC          89.687500    58.437500   35.625000
SA         175.083333   114.750000   62.416667

28. Print the median alcohol consumption per continent for every column

In [56]:
drinks.groupby('continent').agg(
    median_beer=('beer_servings', np.median),
    median_spirit=('spirit_servings', np.median),
    median_wine=('wine_servings', np.median),
)

median_beer  median_spirit  median_wine
continent                                         
AF                32.0            3.0          2.0
AS                17.5           16.0          1.0
EU               219.0          122.0        128.0
OC                52.5           37.0          8.5
SA               162.5          108.5         12.0

29. Print the mean, min and max values for spirit consumption.

In [57]:
drinks.agg(
    mean=('spirit_servings', np.mean),
    min=('spirit_servings', np.min),
    max=('spirit_servings', np.max),
)

spirit_servings
mean        80.994819
min          0.000000
max        438.000000

# Apply: Student Alcohol Consumption
Referencias:
* [apply](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.GroupBy.apply.html?highlight=apply#pandas.core.groupby.GroupBy.apply)

30. Import the dataset from this [address](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/Students_Alcohol_Consumption/student-mat.csv).

In [58]:
url = 'https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/Students_Alcohol_Consumption/student-mat.csv'

## tipos de extensiones típicas datos
# csv -> separados por coma
# tsv -> separados por tabulaciones
# lsv -> separados por espacios
# *csv -> csv importado desde excel puede tener ; de separador
# txt -> puede ser cualquier cosa, suele ser coma


df = pd.read_csv(url, delimiter = ',') # por defecto delimiter = ','

df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...   
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...  \
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

31. Assign it to a variable called df.

32. For the purpose of this exercise slice the dataframe from 'school' until the 'guardian' column

In [59]:
# df.columns
df = df.iloc[:,0:12]

df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob   
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  \
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  
0  course   mother  
1  course   father  
2   other   mother  
3    home   mother  
4    home   father

33. Create a function that will capitalize strings.

In [60]:
def capitalize_string(s):
    return s.capitalize()

34. Capitalize both Mjob and Fjob

In [61]:
df['Mjob'] = df['Mjob'].apply(capitalize_string)
df['Fjob'] = df['Fjob'].apply(capitalize_string)

35. Print the last elements of the data set.

In [62]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob   
0     GP   F   18       U     GT3       A     4     4  At_home   Teacher  \
1     GP   F   17       U     GT3       T     1     1  At_home     Other   
2     GP   F   15       U     LE3       T     1     1  At_home     Other   
3     GP   F   15       U     GT3       T     4     2   Health  Services   
4     GP   F   16       U     GT3       T     3     3    Other     Other   

   reason guardian  
0  course   mother  
1  course   father  
2   other   mother  
3    home   mother  
4    home   father

36. Create a function called majority that returns a boolean value to a new column called legal_drinker (Consider majority as older than 17 years old)

In [63]:
def majority (age):
    return age>17

In [64]:
df['legal_drinker'] = df['age'].apply(majority)

In [65]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob   
0     GP   F   18       U     GT3       A     4     4  At_home   Teacher  \
1     GP   F   17       U     GT3       T     1     1  At_home     Other   
2     GP   F   15       U     LE3       T     1     1  At_home     Other   
3     GP   F   15       U     GT3       T     4     2   Health  Services   
4     GP   F   16       U     GT3       T     3     3    Other     Other   

   reason guardian  legal_drinker  
0  course   mother           True  
1  course   father          False  
2   other   mother          False  
3    home   mother          False  
4    home   father          False

# Merge: Housing Market
Referencias:
* [randint](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.random.randint.html)

This time we will create our own dataset with fictional numbers to describe a house market. As we are going to create random data don't try to reason of the numbers.

37. Import the necessary libraries

38. Create 3 differents Series, each of length 100, as follows: 
<br>
    The first a random number from 1 to 4 
    <br>
    The second a random number from 1 to 3
    <br>
    The third a random number from 10,000 to 30,000

In [66]:
# Genera una Serie aleatoria de números enteros del 1 al 4, de longitud 100
serie1 = pd.Series(np.random.randint(1, 5, size=100))

# Genera una Serie aleatoria de números enteros del 1 al 3, de longitud 100
serie2 = pd.Series(np.random.randint(1, 4, size=100))

# Genera una Serie aleatoria de números enteros del 10000 al 30000, de longitud 100
serie3 = pd.Series(np.random.randint(10000, 30001, size=100))

39. Let's create a DataFrame by joinning the Series by column

In [67]:
hm = pd.DataFrame({'serie1': serie1, 'serie2': serie2, 'serie3': serie3})

hm.head()

serie1  serie2  serie3
0       2       3   29270
1       4       2   13775
2       1       2   27989
3       1       1   13024
4       2       2   23176

40. Change the name of the columns to bedrs, bathrs, price_sqr_meter

In [68]:
hm.rename(columns = {'serie1':'bedrs', 'serie2':'bathrs', 'serie3':'price_sqr_meter'}, inplace = True)

hm.head()

bedrs  bathrs  price_sqr_meter
0      2       3            29270
1      4       2            13775
2      1       2            27989
3      1       1            13024
4      2       2            23176

41. Create a one column DataFrame with the values of the 3 Series and assign it to 'bigcolumn'

In [69]:
bigcolumn = pd.concat([serie1, serie2, serie3], axis=0, ignore_index=True)

# Crea un DataFrame con una sola columna
df = pd.DataFrame(bigcolumn, columns=['bigcolumn'])

42. Oops, it seems it is going only until index 99. Is it true?

In [70]:
df

bigcolumn
0            2
1            4
2            1
3            1
4            2
..         ...
295      23298
296      11799
297      10581
298      25673
299      17169

[300 rows x 1 columns]

43. Reindex the DataFrame so it goes from 0 to 299

In [71]:
df = df.reset_index(drop=True)

df

bigcolumn
0            2
1            4
2            1
3            1
4            2
..         ...
295      23298
296      11799
297      10581
298      25673
299      17169

[300 rows x 1 columns]

# Stats: US - Baby Names

We are going to use a subset of [US Baby Names](https://www.kaggle.com/kaggle/us-baby-names) from Kaggle.  
In the file it will be names from 2004 until 2014

44. Import the dataset from this [address](https://facuzeta.github.io/udesa/01_programacion/guias/datasets/US_Baby_Names_right.csv.zip). 

In [72]:
url = 'https://facuzeta.github.io/udesa/01_programacion/guias/datasets/US_Baby_Names_right.csv.zip'

## tipos de extensiones típicas datos
# csv -> separados por coma
# tsv -> separados por tabulaciones
# lsv -> separados por espacios
# *csv -> csv importado desde excel puede tener ; de separador
# txt -> puede ser cualquier cosa, suele ser coma

df = pd.read_csv(url, compression='zip', delimiter=',') # por defecto delimiter = ','

df.head()

Unnamed: 0     Id     Name  Year Gender State  Count
0       11349  11350     Emma  2004      F    AK     62
1       11350  11351  Madison  2004      F    AK     48
2       11351  11352   Hannah  2004      F    AK     46
3       11352  11353    Grace  2004      F    AK     44
4       11353  11354    Emily  2004      F    AK     41

45. Assign it to a variable called baby_names.

In [73]:
baby_names = df

46. See the first 10 entries

In [74]:
baby_names.head(10)

Unnamed: 0     Id      Name  Year Gender State  Count
0       11349  11350      Emma  2004      F    AK     62
1       11350  11351   Madison  2004      F    AK     48
2       11351  11352    Hannah  2004      F    AK     46
3       11352  11353     Grace  2004      F    AK     44
4       11353  11354     Emily  2004      F    AK     41
5       11354  11355   Abigail  2004      F    AK     37
6       11355  11356    Olivia  2004      F    AK     33
7       11356  11357  Isabella  2004      F    AK     30
8       11357  11358    Alyssa  2004      F    AK     29
9       11358  11359    Sophia  2004      F    AK     28

47. Delete the column 'Unnamed: 0' and 'Id'

In [75]:
baby_names = baby_names.iloc[:,2:]
baby_names.head(10)

Name  Year Gender State  Count
0      Emma  2004      F    AK     62
1   Madison  2004      F    AK     48
2    Hannah  2004      F    AK     46
3     Grace  2004      F    AK     44
4     Emily  2004      F    AK     41
5   Abigail  2004      F    AK     37
6    Olivia  2004      F    AK     33
7  Isabella  2004      F    AK     30
8    Alyssa  2004      F    AK     29
9    Sophia  2004      F    AK     28

48. Is there more male or female names in the dataset?

In [104]:
dfg = baby_names.groupby('Gender').size()
dfg

Gender
F    558846
M    457549
dtype: int64

49. Group the dataset by name and assign to names

In [106]:
baby_names.groupby('Name').size()

Name
Aaban        2
Aadan        4
Aadarsh      1
Aaden      196
Aadhav       1
          ... 
Zyra         7
Zyrah        2
Zyren        1
Zyria       10
Zyriah       9
Length: 17632, dtype: int64

50. How many different names exist in the dataset?

In [108]:
baby_names['Name'].nunique()

17632

51. What is the name with most occurrences?

In [120]:
baby_names.groupby('Name').size().sort_values(ascending=False).head(1)

Name
Riley    1112
dtype: int64

52. How many different names have the least occurrences?

In [125]:
name_counts = baby_names.groupby('Name').size().sort_values(ascending=True)
unique_names = name_counts[name_counts == 1]

len(unique_names)


3682

53. What is the median name occurrence?

In [131]:
name_counts.median()


8.0

54. What is the standard deviation of names?

In [132]:
name_counts.std()

122.0299635081389

55. Get a summary with the mean, min, max, std and quartiles.

In [134]:
name_counts.describe()

count    17632.000000
mean        57.644907
std        122.029964
min          1.000000
25%          2.000000
50%          8.000000
75%         39.000000
max       1112.000000
dtype: float64

# Deleting: Iris
Referencias:
* [fillna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html)

56. Import the dataset from this [address](https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data). 

In [4]:
iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data')

57. Assign it to a variable called iris

In [7]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width        class
0           4.9          3.0           1.4          0.2  Iris-setosa
1           4.7          3.2           1.3          0.2  Iris-setosa
2           4.6          3.1           1.5          0.2  Iris-setosa
3           5.0          3.6           1.4          0.2  Iris-setosa
4           5.4          3.9           1.7          0.4  Iris-setosa

58. Create columns for the dataset

In [6]:
# 1. sepal_length (in cm)
# 2. sepal_width (in cm)
# 3. petal_length (in cm)
# 4. petal_width (in cm)
# 5. class

iris.columns = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'class']


59.  Is there any missing value in the dataframe?

In [8]:
iris.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
class           0
dtype: int64

60.  Lets set the values of the rows 10 to 29 of the column 'petal_length' to NaN

In [9]:
# 60.  Lets set the values of the rows 10 to 29 of the column 'petal_length' to NaN
iris.loc[10:29, 'petal_length'] = np.nan


iris.isnull().sum()

61. Good, now lets substitute the NaN values to 1.0

In [ ]:
iris['petal_length'].fillna(1.0, inplace=True)

62. Now let's delete the column class

63.  Set the first 3 rows as NaN

64.  Delete the rows that have NaN

65. Reset the index so it begins with 0 again

# Bonus:
Si quiere mas ejercicios visitar:
* https://github.com/ajcr/100-pandas-puzzles/blob/master/100-pandas-puzzles.ipynb
* https://www.machinelearningplus.com/python/101-pandas-exercises-python/